### PR TITLE
ISS Location Spice

### PR DESCRIPTION
**Pull request by rishiag**
**What does your Instant Answer do?**
Gives current location of International Space Station

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Nothing really. Just a fun instant answer.

**What is the data source for your Instant Answer? (Provide a link if possible)**
http://open-notify.org/Open-Notify-API/ISS-Location-Now/

**Why did you choose this data source?**
It provides JSON response and data is directly scraped from NASA website

**Are there any other alternative (better) data sources?**
No. Present data source does the job fine

**What are some example queries that trigger this Instant Answer?**
"ISS location", "ISS position", "What is current ISS location", "What is current ISS position"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Space hobbyist, students, etc.

**Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?**
Yes, this one - https://duck.co/ideas/idea/4588/current-location-for-international-space-location

**Which existing Instant Answers will this one supersede/overlap with?**
None that I know of.

**Are you having any problems? Do you need our help with anything?**
Yes, API source suggests it should polled once every 5 second and result should be cached. I was unsure whether 'spice proxy_cache_valid => "200 304 1d";' would work for seconds too, so I did not include it.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
No, API key is not required and there are no costs for API usage.

**Where did you hear about DuckDuckHack? (For first time contributors)**
Hacker News

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
image

## Checklist

[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Ubuntu 12.04-64 Bit
        [x] Google Chrome
        [x] Firefox

### CODE REVIEW

**Comment by jagtalon** [comment]
@rishiag @mintsoft that looks sweet!

**Comment by jagtalon** [comment]
Will review it in a bit. :)

**Comment by bsstoner** [comment]
I added the functionality internally, it's currently waiting to be reviewed & released to production. It will let you pass a zoomLevel override in the meta object (see below). The value is passed through directly to the mapbox/leaflet.js `setZoom` function: http://leafletjs.com/reference.html#map-setzoom The value should be an integer between [4,18], the lower the number the further zoomed out the map (probably just use 4 here).

Also, unrelated to the zoom issue, you need to wrap the Spice.add call in `DDG.require('maps', function(){})` to make sure the mapbox assets get loaded before it tries to add the spice. It's only working right now because the other map IA is also firing on the 'location' keyword and loading the mapbox assets.

```
DDG.require('maps', function(){
  Spice.add({
    id: "isslocation",
    name: "ISSLocation",
    model: 'Place',
    view: 'Map',
    data: [{
        display_name: "International Space Station",
        name: "International Space Station",
        lat: -34.172,
        lon: 156.2613
    }],
    meta: {
        sourceName: "open-notify.org",
        sourceUrl: 'http://open-notify.org',
	zoomLevel: 4
    }
  });
});
```

**Comment by mintsoft** [comment]
@bsstoner is there a PR for that on the duckduckgo repo or is it on the closed-source side?

**Comment by jagtalon** [comment]
Thanks, @bsstoner! @rishiag it's up!

screen shot 2015-01-13 at 1 40 41 pm

@rishiag not sure about the tab name since it's too specific. What about just 'Answer'?

File: share/spice/isslocation/isslocation.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+(function(env) {
+    "use strict";
+    env.ddg_spice_isslocation = function(api_result) {
+        console.log("API result is", api_result);
+        if (api_result.message !== "success") {
```
Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: We need to check if `api_result` exists first.

```javascript
if(!api_result || api_result !== "success") {
    ...
}
```

File: share/spice/isslocation/isslocation.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+(function(env) {
+    "use strict";
+    env.ddg_spice_isslocation = function(api_result) {
+        console.log("API result is", api_result);
```
Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: Remove `console.log` because this will fail on IE 8

File: share/spice/isslocation/isslocation.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        if (api_result.message !== "success") {
+            return Spice.failed('isslocation');
+        }
+        var issLatitude = (Math.round(api_result.iss_position.latitude * 100) / 100).toFixed(2);
+        var issLongitude = (Math.round(api_result.iss_position.longitude * 100) / 100).toFixed(2);
+
+        Spice.add({
```
Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: We might have left it out in the documentation, but we need to wrap it in:

```javascript
DDG.require('maps', function() {
    Spice.add()
}
```

Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: @moollaza @mintsoft probably should be in the docs. :) ^

**Comment by jagtalon** [comment]
@rishiag Clicking on the link doesn't take me anywhere:

screen shot 2015-01-13 at 2 35 48 pm

File: share/spice/isslocation/isslocation.js (hunk 3 above)
Comment (mintsoft) [inline share/spice/isslocation/isslocation.js]: @jagtalon is there several of these or is it just maps?

Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: Yeah it looks like it's just maps. :) Now that I think about it, it should probably pull the data down automatically if we're using the map model kinda like in soundcloud.

File: lib/DDG/Spice/ISSLocation.pm [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+description "International Space Station Location";
+name "ISSLocation";
+source "open-notify";
+code_url "https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/ISSLocation.pm";
+topics "science";
+category "random";
+attribution github => ['https://github.com/rishiag','Rishi Agarwal'],
+           twitter => ['http://twitter.com/rishiagar','Rishi Agarwal'];
+
+triggers any => "iss location", "iss position";
```
Comment (jagtalon) [inline lib/DDG/Spice/ISSLocation.pm]: Other triggers might be "international space station" or maybe "iss tracker / iss tracking"

**Comment by jagtalon** [comment]
@rishiag What do you think of the changes?

File: share/spice/isslocation/isslocation.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            view: 'Map',
+            data: [{
+                display_name: "International Space Station",
+                name: "International Space Station",
+                lat: issLatitude,
+                lon: issLongitude
+            }],
+            meta: {
+                sourceName: "open-notify.org",
+                sourceUrdl: 'http://open-notify.org'
```
Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: Should be `sourceUrl` :)

Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: Adding `zoomLevel: 3` looks about right here

Comment (jagtalon) [inline share/spice/isslocation/isslocation.js]: Actually, `zoomLevel: 1

File: lib/DDG/Spice/ISSLocation.pm (hunk 4 above)
Comment (jagtalon) [inline lib/DDG/Spice/ISSLocation.pm]: Or maybe just "iss" should probably trigger it. :)

Comment (jagtalon) [inline lib/DDG/Spice/ISSLocation.pm]: Also fun ones like, "where is the iss?"

**Comment by moollaza** [comment]
ping! Hey @rishiag , I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this go live!

**Comment by rishiag** [comment]
Hey @moollaza I am really sorry. Actually I was in the process of switching jobs and interviewing with different companies for past month. I was not able to pay any attention. Give me a day or two. Let me setup the environment on my new work laptop and I will give a updated pull request.

**Comment by moollaza** [comment]
>I am really sorry.  Actually I was in the process of switching jobs and interviewing with different companies for past month. I was not able to pay any attention. Give me a day or two. Let me setup the environment on my new work laptop and I will give a updated pull request.

It's absolutely fine! No need to apologize, I was just checking in to make sure you're still interested. Take as much time as you need, there's absolutely no rush :smiley:

**Comment by javathunderman** [comment]
@moollaza Any help required here? Seems like this is moving along well.

**Comment by moollaza** [comment]
@javathunderman it looks like @rishiag may be busy, if you'd like to finish this up I'm sure he wouldn't mind :smiley:

**Comment by javathunderman** [comment]
@jagtalon I might take this one up then, the Forvo IA I was working on wasn’t going anywhere. 😊

**Comment by jagtalon** [comment]
@javathunderman cool then that sounds good! Looking forward to it. :) I can close this when your PR comes along.

**Comment by javathunderman** [comment]
@jagtalon Sorting out those line notes you left. Is there anything else to take care of?

**Comment by mintsoft** [comment]
@javathunderman doesn't look like there's anything else; but do a PR and we'll check it!

**Comment by javathunderman** [comment]
@mintsoft Awesome! Will create a new PR soon.

**Comment by moollaza** [comment]
Continued in #1559

**Pull request closed without merging**